### PR TITLE
Added an algorithm for 2-coloring of bipartite graphs

### PR DESCRIPTION
**Pull request by meghanamreddy**
Added the coloring algorithm for bipartite graphs. Using a simple DFS approach.

### CODE REVIEW

**Comment by AlexandruValeanu** [comment]
@meghanamreddy Doesn't your implementation and [GreedyColoring](https://github.com/jgrapht/jgrapht/blob/master/jgrapht-core/src/main/java/org/jgrapht/alg/color/GreedyColoring.java) behave identically on bipartite graphs?

**Comment by meghanamreddy** [comment]
@AlexandruValeanu To an extent, yes. Given a bipartite graph, my algorithm and GreedyColoring behave identically. However, my implementation is specifically for bipartite graphs using the DFS traversal approach. Can be used to check if a graph is bipartite and also get the 2-coloring. Moreover, this implementation was done as a starting point, hopefully I will be able to extend it to better coloring algorithms.

**Comment by AlexandruValeanu** [comment]
@meghanamreddy 
There are two things I feel like I should mention:

- there is a function for checking whether a graph is bipartite in `GraphTests` (see [link](https://github.com/jgrapht/jgrapht/blob/master/jgrapht-core/src/main/java/org/jgrapht/GraphTests.java#L359)). You might want to check it and benchmark the two implementations (there is no point in having two implementations for the same thing)
- All colouring algorithms implement the `VertexColoringAlgorithm` interface. Your class seems to be compatible with the interface but it doesn't implement. Any good reason for this?

**Comment by d-michail** [comment]
Hi Meghana, 

As @AlexandruValeanu mentioned there is already an implementation which checks whether the graph is bipartite. What is missing (I planned but never had the time to write) is a method which checks if the graph is bipartite and **also** returns the computed partition. 

However, there is no need to mix such an implementation with a coloring algorithm.

Moreover, the use of the `CrossComponentIterator` is highly likely to make the implementation "heavyweight". That iterator is mostly to allow users to be notified on each vertex visit.

**Comment by jsichi** [comment]
Since a bipartite partitioning *is* a 2-coloring, I think it's appropriate to express it as a coloring algorithm (and implement the VertexColoringAlgorithm interface, as Alexandru suggests).  If that adds overhead, then a specialization method for just returning the two sets should be provided, and the overhead cost should be deferred to the coloring-related methods if possible.

But I agree that extending CrossComponentIterator isn't good since the algorithm isn't incremental from the caller's point of view; as Dimitrios says, new iterators should only be published in cases where the iteration needs to be interleaved with the logic of the caller.  (Using an iterator as a private implementation detail of the algorithm is fine, on the other hand, unless it adds unnecessary overhead.)

File: jgrapht-core/src/main/java/org/jgrapht/alg/color/BipartiteColoring.java [hunk 1]
```diff
@@ -0,0 +1,284 @@
+/*
+ * (C) Copyright 2017-2017, by Dimitrios Michail and Contributors.
```
Comment (jsichi) [inline jgrapht-core/src/main/java/org/jgrapht/alg/color/BipartiteColoring.java]: Copyright should be the submitter, and for the current year.

**Comment by meghanamreddy** [comment]
I was not fully aware of the purpose of the iterators. Since I was using a DFS approach, I decided to extend an iterator and work on it. Now that I understand its usage, I realise that adding a new iterator for coloring was not a good idea. 

I would like to express bipartite coloring as a 2-coloring algorithm by implementing the VertexColoringAlgorithm interface as @jsichi suggested. However, as @d-michail suggested, I could also instead add a function which returns the two computed partitions (if the graph is known to be bipartite). With regard to this, my doubt is where should such a function (which returns the partitions of a bipartite graph) be added in the library since adding it to the class GraphTests along with the isBipartite function may not be a good idea. Please let me know which path to take, I will start working on the implementation while ensuring I use lightweight methods.

**Comment by jsichi** [comment]
The existing VertexColoringAlgorithm.getColorClasses can return the two sets.  @d-michail if we add a convenience wrapper method, where do you think we should put it?

GraphTests?  A static method on the algorithm class itself?  Or start a new utility class at top level?

**Comment by d-michail** [comment]
If we decide to have a dedicated class for this, then yes the best location would be in the color package implementing the `VertexColoringAlgorithm`. We still need the `GraphTests.isBipartite()` method. 

Depending on the efficiency of the new implementation (this needs benchmarking with all the classic JVM techniques like running the tests multiple times before measuring performance to trigger hot-swapping, etc.) then we can delegate to the new implementation. 

We can also reuse parts of the current `GraphTests.isBipartite()` method (maybe swap the LinkedList with an ArrayDeque to reduce cache-misses, etc.).

**Comment by meghanamreddy** [comment]
I have started implementing as a class in the color package, I will create a pull request as soon as I finish the implementation. Thank you.

**Comment by meghanamreddy** [comment]
@jsichi @d-michail I have made the changes in the current pull request itself. I have added a class Bipartite2Coloring.java in the color module which implements the VertexColoringAlgorithm interface.

I have extended the code in the isBipartite() method for finding the 2-coloring of a bipartite graph. Hence, it was rather straightforward. The new implementation finds a partition and then colors the vertices. If the input graph is not bipartite, the algorithm throws an exception.

**Comment by d-michail** [comment]
Instead of referencing the old implementation, you should keep the original author's name in the author's list.

Is there some particular reason why you switched from a "queue" to a "stack", or just personal preference?

**Comment by meghanamreddy** [comment]
There is no particular reason for switching the queue with stack. It was just something that I was more comfortable with. And I have added the original authors to the author list.

File: jgrapht-core/src/test/java/org/jgrapht/alg/color/Bipartite2ColoringTest.java [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+import static org.junit.Assert.assertEquals;
+
+/**
+ * Test for the BipartiteColoring class.
+ * 
+ * @author Meghana M Reddy
+ *
+ */
+public class Bipartite2ColoringTest {
+    public void test1() {
```
Comment (jkinable) [inline jgrapht-core/src/test/java/org/jgrapht/alg/color/Bipartite2ColoringTest.java]: Can you add a few lines of text, explaining what you are testing? E.g. "test on non-bipartite graph" "test on disconnected, directed bipartite graph" etc. That makes it easier to assess what kind of test cases you covered without having to draw the graphs.

File: jgrapht-core/src/main/java/org/jgrapht/alg/color/Bipartite2Coloring.java [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+
+            for (E e : graph.edgesOf(v)) {
+                V n = Graphs.getOppositeVertex(graph, e, v);
+                if (unseen.contains(n)) {
+                    stack.addLast(n);
+                    if (!odd.contains(v)) {
+                        odd.add(n);
+                    }
+                } else if (!(odd.contains(v) ^ odd.contains(n))) {
+                    throw new IllegalArgumentException("Input graph is not bipartite");
```
Comment (jkinable) [inline jgrapht-core/src/main/java/org/jgrapht/alg/color/Bipartite2Coloring.java]: I would prefer to return 'null', indicating that a 2-coloring does not exist. That is more graceful than returning a runtime exception. This behavior should obviously be documented.

**Comment by jkinable** [comment]
This implementation is obvious almost identical to the `isBipartiteGraph` test in `GraphTests`. I would therefore recommend to move the implementation of the isBipartiteGraph test from GraphTests to this class (similar to isEulerian in GraphTest). This prevents code duplication.

File: jgrapht-core/src/test/java/org/jgrapht/alg/color/Bipartite2ColoringTest.java (hunk 2 above)
Comment (meghanamreddy) [reply jgrapht-core/src/test/java/org/jgrapht/alg/color/Bipartite2ColoringTest.java]: I have added additional test cases and their corresponding descriptions.

**Comment by meghanamreddy** [comment]
I have moved the implementation of `isBipartite` to `Bipartite2Coloring` and used this implementation for `isBipartite` method, as suggested.

File: jgrapht-core/src/main/java/org/jgrapht/alg/color/Bipartite2Coloring.java [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+     */
+    public Bipartite2Coloring(Graph<V, E> graph)
+    {
+        this.graph = Objects.requireNonNull(graph, "Graph cannot be null");
+    }
+    
+    /**
+     * {@inheritDoc}
+     * If the input graph is not bipartite, the Map object returned
+     * by the getColors() method is null.
```
Comment (jsichi) [inline jgrapht-core/src/main/java/org/jgrapht/alg/color/Bipartite2Coloring.java]: Wouldn't it be cleaner to have getColoring() itself return null?

File: jgrapht-core/src/main/java/org/jgrapht/alg/color/Bipartite2Coloring.java [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+ * @param <V> the graph vertex type
+ * @param <E> the graph edge type
+ */
+public class Bipartite2Coloring<V, E> 
+    implements VertexColoringAlgorithm<V>
+{
+    /**
+     * The input graph
+     */
+    protected final Graph<V, E> graph;
```
Comment (jsichi) [inline jgrapht-core/src/main/java/org/jgrapht/alg/color/Bipartite2Coloring.java]: Why is this protected?

File: jgrapht-core/src/main/java/org/jgrapht/alg/color/Bipartite2Coloring.java [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+    
+    /**
+     * Stores the final colors of the vertices after finding a 2-coloring.
+     */
+    Map<V, Integer> bipartiteColors = new HashMap<>();
```
Comment (jsichi) [inline jgrapht-core/src/main/java/org/jgrapht/alg/color/Bipartite2Coloring.java]: Why isn't this private?

**Comment by jkinable** [comment]
@meghanamreddy can you address @jsichi comments?

**Comment by jkinable** [comment]
@meghanamreddy can you make the requested modifications such that we can finish this PR? Otherwise we'll have to tag this PR as abandoned which would be pity.

**Comment by jkinable** [comment]
This class is no longer needed due to the merge of #613 which implements a class to identify the partitions of a bipartite graph.

**Pull request closed without merging**